### PR TITLE
Generate AppImage for Linux

### PR DESCRIPTION
**Pull request by probonopd**
Someone from the atom team please see below, solve the remaining issue and then merge this once it builds "green".

This PR, when merged, will compile this application on [Travis CI](https://travis-ci.org/) upon each `git push`, and upload an [AppImage](http://appimage.org/).

It tries to mirror 1:1 what the debian pipeline is doing, using the same binaries. So __this should _not_ add additional burden to the Atom team__. It is merely a self-mounting filesystem image. Think of it as putting the contents of the deb into a self-mounting ISO.

Providing an [AppImage](http://appimage.org/) would have, among others, these advantages:
- Applications packaged as an AppImage can run on many distributions (including Ubuntu, Fedora, openSUSE, CentOS, elementaryOS, Linux Mint, and others)
- One app = one file = super simple for users: just download one AppImage file, [make it executable](http://discourse.appimage.org/t/how-to-make-an-appimage-executable/80), and run
- No unpacking or installation necessary
- No root needed
- No system libraries changed
- Works out of the box, no installation of runtimes needed
- Optional desktop integration with `appimaged`
- Optional binary delta updates, e.g., for continuous builds (only download the binary diff) using AppImageUpdate
- Can optionally GPG2-sign your AppImages (inside the file)
- Works on Live ISOs
- Can use the same AppImages when dual-booting multiple distributions

[Here is an overview](https://github.com/probonopd/AppImageKit/wiki/AppImages) of projects that are already distributing upstream-provided, official AppImages.

__I need help on the following issue:__  What translates the lowercase `--create-debian-package` to make `if (argv.createDebianPackage)` in `script/build` trigger? I _think_ I have added and am calling `--create-appimage` but I am currently getting

> Skipping creating AppImage. Specify the --create-appimage option to create it.

Any help greatly appreciated.

AppImage developers are on #AppImage on irc.freenode.net.

### CODE REVIEW

**Comment by JonasT** [comment]
Maybe it would be better to provide this as a .zip which has a .home directory added? Otherwise people might lose their settings or inadvertently break those of an existing system-wide install when running it.. (unless the AppImage can somehow be forced to always use a local neighbor folder as `$HOME` and not only if the user pre-created it)

**Comment by probonopd** [comment]
@JonasT a `.zip` provides no advantage over an AppImage. You can unpack an AppImage very easily by running ``--appimage-extract`. The AppImage is an _alternative_ to a system-wide install (e.g, think of someone without root rights who cannot do a system-wide install at all).

**Comment by JonasT** [comment]
@probonopd I think I might not have been clear. The .zip should contain the AppImage in packed form but also an `<AppImage>.home` folder. Alternatively, I propose that portable Atom should be packaged with a startup script as suggested by you here https://github.com/AppImage/AppImageSpec/issues/4#issuecomment-321091078 which ensures settings are never stored in the regular actual user home.

**Comment by probonopd** [comment]
@JonasT please [_do not_ ](https://github.com/AppImage/AppImageKit/wiki/Creating-AppImages#common-mistake)put AppImages into other archives such as zip files as it would entirely defeat the purpose of the format.

**Comment by JonasT** [comment]
I was simply pointing out a potential culprit when the `$HOME` redirection for portable settings needs to be somehow proactively activated by the user. I don't have any strong opinions on how specifically to solve it

**Comment by hultberg** [comment]
@probonopd Try this for checking `--create-appimage`:
```js
if (argv.createAppimage) {
```

There is only uppercase when there is a dash separating words. Like: `--create-debian-package`: `argv.createDebianPackage`

**Comment by probonopd** [comment]
Thanks @hultberg. Would never have tried this...

**Comment by probonopd** [comment]
Still running into issues,

```
  console.log(`Copying icon into "${appimagePackageIconsDirPath}"`)
  fs.copySync(
    path.join(packagedAppPath, 'resources', 'app.asar.unpacked', 'resources', 'atom.png'),
    path.join(appimagePackageIconsDirPath, `${atomExecutableName}.png`)
  )
```

results in

```
Error: ENOENT: no such file or directory, stat '/home/travis/build/probonopd/atom/out/atom-1.21.0-dev-e792504-amd64/resources/app.asar.unpacked/resources/atom.png'
```

Now, I believe to have copied over that code 1:1 from the debian script...

**Comment by probonopd** [comment]
Can someone take this from here?

File: script/lib/create-appimage.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  const apmExecutableName = CONFIG.channel === 'beta' ? 'apm-beta' : 'apm'
+  const appName = CONFIG.channel === 'beta' ? 'Atom Beta' : 'Atom'
+  const appDescription = CONFIG.appMetadata.description
+  const appVersion = CONFIG.appMetadata.version
+  let arch
+  if (process.arch === 'ia32') {
+    arch = 'i386'
+  } else if (process.arch === 'x64') {
+    arch = 'amd64'
+  } else if (process.arch === 'ppc') {
```
Comment (damieng) [inline script/lib/create-appimage.js]: PowerPC is not a supported platform for Atom.

**Comment by damieng** [comment]
Our position has not changed since https://github.com/atom/atom/pull/13921

**Comment by Holyprogrammer** [comment]
@probonopd Maybe [vscode](https://github.com/Microsoft/vscode) will welcome appimages better. Although I daresay that atom and vscode is ultimately owned by the same company. Was really looking forward to this :(

**Comment by probonopd** [comment]
Why do you think a feature request to "Generate AppImage for Linux" is off-topic @Holyprogrammer?

Your comment really leaves me puzzled:

> Was really looking forward to this :(

So why are you marking this as off-topic then, instead of lobbying for it?

**Comment by Holyprogrammer** [comment]
> Why do you think a feature request to "Generate AppImage for Linux" is off-topic @Holyprogrammer?
> 
> Your comment really leaves me puzzled:
> 
> > Was really looking forward to this :(
> 
> So why are you marking this as off-topic then, instead of lobbying for it?

I didn't mark it as off topic. Maybe some mod did it

**Comment by Holyprogrammer** [comment]
@probonopd I didn't mark it as off-topic. Can project owners do it too?

**Pull request closed without merging**